pyzim/pyzim.pyx

```diff
@@ -1,338 +1,183 @@
-        return self.c_zim_article.mimeType.decode('UTF-8')
-
-    @mimetype.setter
-    def mimetype(self, new_mimetype):
-        """Set the article's mimetype"""
-        self.c_zim_article.mimeType = new_mimetype.encode('UTF-8')
-
+        return self.c_article.getMimeType().decode('UTF-8')
+    
     @property
-    def is_redirect(self):
-        """Get if the article is a redirect"""
-        return self.c_zim_article.isRedirect()
+    def content(self):
+        blob = self.c_article.getData()
+        return blob.data()[:blob.size()].decode('UTF-8')
 
+    # This changes with implementation
     @property
     def can_write(self):
-        """Get if the article is valid for writing"""
-        # An article must have at least non empty url to be writable
-        # Content can be empty if article is a redirect
-        if self.longurl and (self.content or self.is_redirect):
-            self._can_write = True
-        else:
-            self._can_write = False
-        return self._can_write
+        raise NotImplementedError
 
-#########################
-#       ZimReader       #
-#########################
 
-cdef class ZimReader:
-    cdef zim.File *c_file
-    cdef zim.ZimSearch *c_search
-    cdef object _filename
-
-    _metadata_keys =[
-        "Name", 
-        "Title", 
-        "Creator",
-        "Publisher",
-        "Date",
-        "Description",
-        "Language",
-        # Optional
-        "LongDescription",
-        "Licence",
-        "Tags",
-        "Flavour",
-        "Source",
-        "Counter",
-        "Scraper"]
 
-    def __cinit__(self, str filename):
-        self.c_file = new zim.File(filename.encode('UTF-8'))
-        self.c_search = new zim.ZimSearch(self.c_file)
-
-        self._filename = self.c_file.getFilename().decode("UTF-8", "strict")
+#------- ZimArticle pure virtual methods --------
+cdef public api:
+    string string_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning a string"""
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        try:
+            func = getattr(art, method.decode('UTF-8'))
+        except AttributeError:
+            error[0] = 1
+            raise 
+        else:
+            error[0] = 0
+            value = func()
+            return value.encode('UTF-8')
+    
+    Blob blob_cy_call_fct(void *ptr, string method, int *error):
+        """Lookup and execute a pure virtual method on ZimArticle returning a Blob"""
+        cdef ZimArticle art = <ZimArticle>(ptr)
+        cdef ZimBlob blob = ZimBlob(b'\x00')
+        try:
+            func = getattr(art, method.decode('UTF-8'))
+        except AttributeError:
+            error[0] = 1
+            raise 
+        else:
+            error[0] = 0
+            blob = func()
+            return dereference(blob.c_blob) 
 
-    def __dealloc__(self):
-        if self.c_file != NULL:
-            del self.c_file
+cdef public api:    
```

Comment:
Why a second `cdef public api:` ?

Those functions are mainly utility function used by our c++ code.
Even if not necessary, it may be a good thing to move them in a different `pyx` file.

Reply:
Can be moved. ✅